[BEAM-12378] GroupIntoBatches improvements

reuvenlax opened the pull request:
A few improvements to GroupIntoBatches

1. Add GroupIntoBaches.ofByteSize, allowing the user to set the target byte size for a batch.
2. Use OnWindowExpiration instead of setting an explicit end-of-window timer.
3. Use injected key parameter instead of separately storing the key in state.

reuvenlax commented:
R: @nehsyc

reuvenlax commented:
Run Java PreCommit

reuvenlax commented:
@anantdamle I'm not completely sure I understand the use case. One comment I will make on the implementation, is that the existing GroupIntoBatches is very careful to mostly append to the BagState, and only to read from the BagState when it's ready to flush it. Your version reads the bag on every element, which can be very expensive (in fact it is O(n^2) in the number of elements).

anantdamle commented:
@reuvenlax thanks for the heads up on time-complexity of reading BagState (I honestly didn't know). 
Can you help suggest the best way to handle the following scenario:

Context: I need to represent a nested-repeated data as flat-tables.
e.g.
lets say I have two input records, that need to be accumulated.
<table>
<thead>
<th>record-1</th>
<th>record-2</th>
<tbody>
<td><pre>{ id: 1, num: 1.23, arr: ["a", "b"] }</pre></td>
<td><pre>{ id: 2, num: 4.56, arr: [ "d", "e", "f"] }</pre></td>
</tbody>
</table>

The output is, the size of accumulated batch is actually not the sum of serialized size of individual elements, instead, the accumulator needs access to at least the headers list to compute the effective size of the accumulated batch.
<table>
<thead>
<th>id</th>
<th>num</th>
<th>arr[0]</th>
<th>arr[1]</th>
<th>arr[2]</th>
</thead>
<tbody>
<tr>
<td>1</td>
<td>1.23</td>
<td>"a"</td>
<td>"a"</td>
<td>[empty]</td>
</tr>
<tr>
<td>2</td>
<td>4.56</td>
<td>"d"</td>
<td>"e"</td>
<td>"f"</td>
</tr>
</tbody>
</table>

reuvenlax commented:
@anantdamle the time complexity of reading and iterating a bag is simply the size of the bag. However if you do that on every element as the bag grows the cost will be 1 + 2 + 3 + ... + n = O(n^2)

I'm not sure I understand your use case.- are you trying to unroll the loop into the record? This might be better discussed on user@ or dev@

nehsyc commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/GroupIntoBatchesTest.java:
```diff
@@ -126,6 +127,99 @@ public Void apply(Iterable<KV<String, Iterable<String>>> input) {
+            .apply(
+                GroupIntoBatches.ofByteSize(
+                    BATCH_SIZE_BYTES,
+                    s -> {
+                      try {
+                        return StringUtf8Coder.of().getEncodedElementByteSize(s);
+                      } catch (Exception e) {
+                        throw new RuntimeException(e);
+                      }
+                    }))
```
Maybe use a different weigher than the default?

reuvenlax replied:
ddone

nehsyc reviewed (commented):
Thanks!

reuvenlax commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupIntoBatches.java:
```diff
@@ -101,29 +111,69 @@
+  public abstract static class BatchingParams<InputT> implements Serializable {
+    public static <InputT> BatchingParams<InputT> create(
+        long batchSize,
+        long batchSizeByes,
```
done

reuvenlax commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupIntoBatches.java:
```diff
@@ -101,29 +111,69 @@
-  public abstract static class BatchingParams implements Serializable {
-    public static BatchingParams create(long batchSize, Duration maxBufferingDuration) {
-      return new AutoValue_GroupIntoBatches_BatchingParams(batchSize, maxBufferingDuration);
+  public abstract static class BatchingParams<InputT> implements Serializable {
```
added

reuvenlax commented:
@nehsyc addressed comments. Also added Dataflow batch translation for the new code path (Iterables.partition does make a copy of each subpartition, so I don't think it has any performance advantage here).

reuvenlax commented:
Run Dataflow ValidatesRunner

nehsyc commented on runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/GroupIntoBatchesOverride.java:
```diff
@@ -87,14 +94,22 @@ private BatchGroupIntoBatches(long batchSize) {
+                          currentBatch.clear();
+                          batchSizeBytes = 0;
+                        }
                       }
```
Should we also emit `currentBatch` (the remaining partial batch) after the loop?

reuvenlax replied:
Yes - I noticed that as well and pushed a fix.

nehsyc commented on runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowRunnerTest.java:
```diff
@@ -1728,24 +1733,93 @@ private void verifyGroupIntoBatchesOverride(
+    if (withShardedKey) {
+      output =
+          input
+              .apply(GroupIntoBatches.<String, String>ofSize(batchSizeBytes).withShardedKey())
```
ofByteSize(batchSizeBytes)

reuvenlax replied:
done

reuvenlax commented:
Run Dataflow ValidatesRunner

reuvenlax commented:
run dataflow validatesrunner

nehsyc reviewed (approved):
Looks good!

reuvenlax commented:
Run Java PreCommit

reuvenlax commented:
Run Java PreCommit

reuvenlax commented:
Run Dataflow ValidatesRunner

reuvenlax commented:
Run Dataflow ValidatesRunner

reuvenlax commented:
Note: Java precommit completed successfully already. Current failure is cause by a known flaky test

steveniemitz commented on runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/GroupIntoBatchesOverride.java:
```diff
@@ -87,14 +94,25 @@ private BatchGroupIntoBatches(long batchSize) {
+                          c.output(KV.of(c.element().getKey(), currentBatch));
+                          // Call clear() since that allows us to reuse the array memory for
+                          // subsequent batches.
+                          currentBatch.clear();
```
looking at this, I don't understand how this is safe.  Isn't this mutating an element once it's been emitted? If fused step happens to store this list somewhere outside of the processElement call (state, buffer between start/end bundle, etc), it'll be mutated out from under it.

Is there a copy happening somewhere else under the covers that makes this safe?

reuvenlax replied:
oof, this is a bug.

It's probably ok in many cases, since the fused fn will be executed immediately when output is called. However the fused fn is allowed to store the element in memory and process it in finishBundle, and if that happens the data will be stomped. We need to remove this buggy optimization :P
